subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/BuildSrcDeprecatedInSettingsIntegrationTest.groovy

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class BuildSrcDeprecatedInSettingsIntegrationTest extends AbstractIntegrationSpec {
+
+    def "buildSrc classes used in settings is deprecated"() {
+        file('buildSrc/build.gradle') << """
+            apply plugin: 'groovy'
+            
+            repositories {
+                mavenCentral()
+            }
+
+            dependencies {  
+                compile 'org.apache.commons:commons-math3:3.6.1'                
+            }
+        """
+
+        file('buildSrc/src/main/java/org/acme/build/SomeBuildSrcClass.java') << """
+        package org.acme.build;
+        
+        public class SomeBuildSrcClass {
+            public static void foo(String foo){
+                System.out.println(foo);
+            }
+        }
+        """
+        file('buildSrc/src/main/java/org/acme/build/SomeOtherBuildSrcClass.java') << """
+        package org.acme.build;
+        
+        public class SomeOtherBuildSrcClass {
+            public static void foo(String foo){
+                System.out.println(foo);
+            }
+        }
+        """
+        settingsFile << """
+            buildscript {
+                repositories {
+                    mavenCentral()
+                }
+
+                dependencies {  
+                    classpath 'org.apache.commons:commons-lang3:3.9'                
+                }
+            }
+
+            org.acme.build.SomeBuildSrcClass.foo("from settings.gradle")
```

Comment:
This is very clever.

I think we should also add something explicitly to catch plugins that come from buildSrc and are applied in settings because the deprecation would otherwise be pretty strange.

Reply:
I looked into this today and find that kind of tricky. the only thing I can think of is parsing the resource name and provide a specific warning when it is `META-INF/gradle-plugins/X.properties`. See my other reply regarding a general more generic deprecation warning here instead